File: backend/src/main/kotlin/com/example/backend/controllers/InventaireController.kt
package com.example.backend.controllers

import com.example.backend.dtos.InventaireNewCreatetDto
import com.example.backend.dtos.InventaireOneProductUpdateRequestDto
import com.example.backend.dtos.InventaireRequestDto
import com.example.backend.dtos.InventaireUpdateRequestDto
import com.example.backend.models.Inventaire
import com.example.backend.models.ProduitInventaire
import com.example.backend.models.User
import com.example.backend.services.InventaireService
import com.example.backend.services.UserService
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.web.bind.annotation.*

@RestController
@RequestMapping("/api/admin/inventaire")
class InventaireController(
  private val inventaireService: InventaireService
) {

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/create")
  fun creerInventaire(
    @RequestBody data: InventaireRequestDto
  ): ResponseEntity<Inventaire> {
    val inventaire = inventaireService.creerInventaire(data)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PostMapping("/create-new")
  fun creerInventaireNew(
    @RequestBody data: InventaireNewCreatetDto
  ): ResponseEntity<Inventaire> {
    val inventaire = inventaireService.creerInventaireNew(data)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/close/{id}")
  fun cloturerInventaire(@PathVariable id: Long): ResponseEntity<Inventaire> {
    val inventaire = inventaireService.cloturerInventaire(id)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PutMapping("/update/{id}")
  fun mettreAJourInventaire(
    @RequestBody data: InventaireUpdateRequestDto
  ): ResponseEntity<Inventaire> {
    val inventaire = inventaireService.mettreAJourInventaire(data)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @PutMapping("/update/valid/product/{id}")
  fun addProductToInventory(
    @RequestBody data: InventaireOneProductUpdateRequestDto
  ): ResponseEntity<ProduitInventaire> {
    val inventaire = inventaireService.valideProductToInventory(data)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @DeleteMapping("/update/invalid/product/{id}")
  fun invalideProductToInventory(
    @PathVariable id: String
  ): ResponseEntity<Map<String, Any?>> {
    val inventaire = inventaireService.invalideProductToInventory(id)
    return ResponseEntity.ok(inventaire)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/list")
  fun listerInventaires(
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "asc") direction: String
  ): ResponseEntity<Page<Inventaire>> {
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.fromString(direction), sort))
    val inventaires = inventaireService.listerInventaires(pageable)
    return ResponseEntity.ok(inventaires)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/list/new")
  fun listerInventairesCustom(
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "asc") direction: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.fromString(direction), sort))
    val inventaires = inventaireService.listerInventairesCustom(pageable)
    return ResponseEntity.ok(inventaires)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/{id}/products")
  fun listerProduitsParInventaire(
    @PathVariable id: Long,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "asc") direction: String
  ): ResponseEntity<Page<ProduitInventaire>> {
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.fromString(direction), sort))
    val produits = inventaireService.listerProduitsParInventaire(id, pageable)
    return ResponseEntity.ok(produits)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/pageable/{id}/products")
  fun listerProduitsParInventaireAsMap(
    @PathVariable id: String,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(required = false) search: String?,
    @RequestParam(defaultValue = "desc") direction: String
//    @RequestParam(defaultValue = "asc") direction: String
  ): ResponseEntity<Page<Map<String, Any?>>> {

    var newSort = if (direction.equals("asc", ignoreCase = true)) {
      Sort.by(Sort.Direction.ASC, sort)
    } else {
      Sort.by(Sort.Direction.DESC, sort)
    }
    val pageable = PageRequest.of(page, size, newSort)
    val produits = inventaireService.listerProduitsParInventaireAsMap(search,id, pageable)
    return ResponseEntity.ok(produits)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/pageable/{id}/filter/products")
  fun listerProduitsParInventaireAvecFiltre(
    @PathVariable id: String,
    @RequestParam(defaultValue = "equal") filtre: String,
    @RequestParam(defaultValue = "0") page: Int,
    @RequestParam(defaultValue = "10") size: Int,
    @RequestParam(defaultValue = "id") sort: String,
    @RequestParam(defaultValue = "desc") direction: String
//    @RequestParam(defaultValue = "asc") direction: String
  ): ResponseEntity<Page<Map<String, Any?>>> {
//    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "dateDebut"))
    val pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.fromString(direction), sort))
    val produits = inventaireService.listerProduitsParInventaireAvecFiltre(id, pageable,filtre)
    return ResponseEntity.ok(produits)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/info/produit_inventaire/{id}")
  fun getInfoProduitsInventaire(
    @PathVariable id: String
  ): ResponseEntity<Map<String, Any?>> {
    val produits = inventaireService.getInfoProduitsInventaire(id)
    return ResponseEntity.ok(produits)
  }

  @CrossOrigin(origins = ["http://localhost:4200"])
  @PreAuthorize("isAuthenticated()")
  @GetMapping("/terminer/{id}")
  fun terminerInventaire(@PathVariable id: Long,
                         @RequestParam commentaire: String): ResponseEntity<Inventaire> {
    val inventaire = inventaireService.terminerInventaire(id,commentaire)
    return ResponseEntity.ok(inventaire)
  }

}

File: backend_py/pharmaPython/app/api/v1/inventaire_controller.py
from fastapi import APIRouter, Depends, Query, HTTPException
from sqlalchemy.orm import Session
from typing import List, Dict, Any, Optional

from app.api.deps import get_db
from app.models.employe import Employe
from app.models.inventaire import InventaireSchema
from app.models.produit_inventaire import ProduitInventaireSchema
from app.schemas.inventaire_dto import InventaireRequestDto, InventaireNewCreatetDto, \
  InventaireOneProductUpdateRequestDto, InventaireUpdateRequestDto
from app.services.inventaire_service import InventaireService
from app.utility.jwt_authentication import jwt_authentication
from app.utility.user_utils import UserUtils

router = APIRouter(
  prefix="/admin/inventaire",
  tags=["Inventaire"],
  dependencies=[Depends(jwt_authentication)],
)


@router.post("/create", response_model=InventaireSchema)
def creer_inventaire(data: InventaireRequestDto, db: Session = Depends(get_db)):
  return InventaireService(db).creer_inventaire(data)


@router.post("/create-new")
def creer_inventaire_new(data: InventaireNewCreatetDto, db: Session = Depends(get_db),
                         employe: Employe = Depends(UserUtils.get_current_employe)):
  return InventaireService(db).creer_inventaire_new(data, employe)


@router.get("/close/{id}")
def cloturer_inventaire(id: int, db: Session = Depends(get_db)):
  return InventaireService(db).cloturer_inventaire(id)


@router.put("/update/{id}", response_model=InventaireSchema)
def mettre_a_jour_inventaire(data: InventaireUpdateRequestDto, db: Session = Depends(get_db),
                             employe: Employe = Depends(UserUtils.get_current_employe)):
  return InventaireService(db).mettre_a_jour_inventaire(data, employe)


@router.put("/update/valid/product/{id}", response_model=ProduitInventaireSchema)
def add_product_to_inventory(data: InventaireOneProductUpdateRequestDto, db: Session = Depends(get_db),
                             employe: Employe = Depends(UserUtils.get_current_employe)):
  return InventaireService(db).valide_product_to_inventory(data, employe)


@router.delete("/update/invalid/product/{id}", response_model=Dict[str, Any])
def invalide_product_to_inventory(id: str, db: Session = Depends(get_db)):
  return InventaireService(db).invalide_product_to_inventory(id)


@router.get("/list")
def lister_inventaires(
  page: int = Query(0), size: int = Query(10),
  sort: str = Query("id"), direction: str = Query("desc"),
  db: Session = Depends(get_db),
):
  return InventaireService(db).lister_inventaires_custom(page, size, sort, direction)


@router.get("/list/new")
def lister_inventaires_custom(
  page: int = Query(0), size: int = Query(10),
  sort: str = Query("id"), direction: str = Query("desc"),
  db: Session = Depends(get_db),
):
  return InventaireService(db).lister_inventaires_custom(page, size, sort, direction)


@router.get("/{id}/products")
def lister_produits_par_inventaire(
  id: int, page: int = 0, size: int = 10, sort: str = "id", direction: str = "desc",
  db: Session = Depends(get_db),
):
  return InventaireService(db).lister_produits_par_inventaire(id, page, size, sort, direction)


@router.get("/pageable/{id}/products")
def lister_produits_par_inventaire_as_map(
  id: str, search: Optional[str] = None,
  page: int = 0, size: int = 10, sort: str = "id", direction: str = "desc",
  db: Session = Depends(get_db),
):
  return InventaireService(db).lister_produits_par_inventaire_as_map(search, id, page, size, sort, direction)


@router.get("/pageable/{id}/filter/products")
def lister_produits_par_inventaire_avec_filtre(
  id: str, filtre: str = "equal",
  page: int = 0, size: int = 10, sort: str = "id", direction: str = "desc",
  db: Session = Depends(get_db),
):
  return InventaireService(db).lister_produits_par_inventaire_avec_filtre(id, page, size, sort, direction, filtre)


@router.get("/info/produit_inventaire/{id}")
def get_info_produits_inventaire(id: str, db: Session = Depends(get_db)):
  return InventaireService(db).get_info_produits_inventaire(id)


@router.get("/terminer/{id}", response_model=InventaireSchema)
def terminer_inventaire(id: int, commentaire: str, db: Session = Depends(get_db)):
  return InventaireService(db).terminer_inventaire(id, commentaire)
